### ISSUE TITLE
Storage: HMAC tests not cleaning up after themselves. HMAC keys limit reached.

### ISSUE THREAD

**Issue by amanda-tarafa** [GoogleCloudPlatform/dotnet-docs-samples#1272]
Seems like HMAC tests are not properly deleting HMAC keys and we have reached limit. All HMAC keys related tests are failing.

Sample [CI output](https://source.cloud.google.com/results/invocations/00fb02fd-8e97-40af-b459-c95f324ef2dd/targets/github%2Fdotnet-docs-samples%2Fstorage%2Fapi%2FStorage.Samples.Tests%2FTestResults/tests).

**Comment by amanda-tarafa** [comment]
Actually all the tests are cleaning up after themselves. This seems to be because of sporadic failures over time in deleting the keys, and they have build up. I will clear the existing keys but keep this issue open for a while to see how soon we see it build back up.

**Comment by amanda-tarafa** [comment]
I just deleted all HMAC keys. Let's give it a couple of weeks and see how soon it builds of egain.

**Comment by amanda-tarafa** [comment]
And we have now [an error](https://source.cloud.google.com/results/invocations/24ac2550-b9b4-4959-b545-9a657e227787/targets/github%2Fdotnet-docs-samples%2Fstorage%2Fapi%2FStorage.Samples.Tests%2FTestResults/tests) which seems because we are trying to clean an HMAC key that was already delete.

**Comment by amanda-tarafa** [comment]
We've reached limit [again](https://source.cloud.google.com/results/invocations/4d68d111-4399-4c1f-8037-daf4c2eadec9/targets/github%2Fdotnet-docs-samples%2Fstorage%2Fapi%2FStorage.Samples.Tests%2FTestResults/tests).

**Comment by amanda-tarafa** [comment]
@tritone Assigning to you as well. There's some issue with the HMAC cleanup, it's there, but it might not be working always. HMAC keys are filling up quite quickly.

Assigning to you as well so that the first one of us with badnwidth can take a look. Thanks

**Comment by tritone** [comment]
I did a little bit of investigation around this.

I see the code to clean up the test keys is [here](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/2b8c5d9fc13250621793453106c76b386e948498/storage/api/Storage.Samples.Tests/BucketFixture.cs#L121) and called in tests for example [here](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/master/storage/api/Storage.Samples.Tests/ActivateHmacKeyTest.cs#L47). The cleanup code actually uses the deactivate/delete samples rather than using the client directly.

I'm a little unsure about how error handling works in dotnet-- @amanda-tarafa do you have a sense of what would happen if there is a failure in the test or cleanup code? Would it cause the test to fail, or just pass unnoticed?  It looks like if the test itself fails, the cleanup definitely wouldn't happen-- is there a way to change this? (We used pytest fixture teardown in Python and`defer` in Go, both of which make cleanup happen even if the test has a fatal error in execution.)

One other thought is that, if we don't expect multiple tests runs to be happening in parallel using the same testing project, we could actually just add cleanup code to delete all old keys on the project before every test run. We implemented this in Go [here](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/storage/service_account/hmac/hmac_test.go#L44). These keys should not be used by anything else in the project setup so there isn't much risk there.

**Comment by amanda-tarafa** [comment]
Reopening, there's a flake leftover, I'll look into it:

```
[xUnit.net 00:00:08.19]     ActivateHmacKeyTest.TestActivateHmacKey [FAIL]
  X ActivateHmacKeyTest.TestActivateHmacKey [2s 21ms]
  Error Message:
   Google.GoogleApiException : Google.Apis.Requests.RequestError
Access ID not found in project. [404]
Errors [
	Message[Access ID not found in project.] Location[ - ] Reason[notFound] Domain[global]
]

  Stack Trace:
     at Google.Apis.Requests.ClientServiceRequest`1.ParseResponse(HttpResponseMessage response)
   at Google.Apis.Requests.ClientServiceRequest`1.Execute()
   at Google.Cloud.Storage.V1.StorageClientImpl.GetHmacKey(String projectId, String accessId, GetHmacKeyOptions options)
   at DeactivateHmacKeySample.DeactivateHmacKey(String projectId, String accessId) in /tmpfs/src/github/dotnet-docs-samples/storage/api/Storage.Samples/DeactivateHmacKey.cs:line 28
   at ActivateHmacKeyTest.TestActivateHmacKey() in /tmpfs/src/github/dotnet-docs-samples/storage/api/Storage.Samples.Tests/ActivateHmacKeyTest.cs:line 37
```

**Comment by amanda-tarafa** [comment]
It seems that, even though the key has been successfully created, in some cases it is not inmediately available. @tritone is there a way to confirm this? Because if this is the case, we should add a sleep on samples, not just on tests, since this would be something that clients themselves will encounter.

**Comment by amanda-tarafa** [comment]
Closing in favour of #1367. After #1339, we are not reaching quota anymore. But we see an ocasional flake as described in #1367
